hail/src/main/scala/is/hail/expr/ir/AggOp.scala

```diff
@@ -4,30 +4,72 @@ import is.hail.expr.ir.agg.Extract
 import is.hail.expr.types.physical.PType
 import is.hail.expr.types.virtual._
 
+object AggStateSignature {
+  def apply(sig: AggSignature): AggStateSignature = AggStateSignature(Map(sig.op -> sig), sig.op)
+}
+
+case class AggStateSignature(m: Map[AggOp, AggSignature], default: AggOp, nested: Option[Seq[AggStateSignature]] = None) {
+  lazy val defaultSignature: AggSignature = m(default)
+  lazy val resultType: Type = Extract.getResultType(this)
+  def lookup(op: AggOp): AggSignature = m(op)
+
+  def toCanonicalPhysical: AggStatePhysicalSignature = AggStatePhysicalSignature(
```

Comment:
One thing we didn't discuss in much detail is how AggStateSignature non-canonical physical types should be  handled (in the lower to CollectDistributedArray world). This feels like a step back relative to what we've done with virtual types (remove the 1:1 mapping), though as I understand it a necessary, temporary step.

Reply:
I have the code to do this, but wanted to break it up into two PRs!